File: app/public/generate-icons.py
#!/usr/bin/env python3
"""
Generatore Icone PWA - Sistema Gestione ISO 9001:2015

Genera tutte le icone richieste dal manifest.json utilizzando Pillow.
Crea un'icona base con logo QS Studio e la ridimensiona per tutte le risoluzioni.

Requisiti: pip install pillow

Standard: ISO 9001:2015 punto 7.5 (documented information)
"""

from PIL import Image, ImageDraw, ImageFont
import os

# Configurazione colori QS Studio
PRIMARY_COLOR = "#1976d2"  # Blue ISO
BACKGROUND_COLOR = "#ffffff"  # White
TEXT_COLOR = "#ffffff"  # White text

# Dimensioni richieste
SIZES = [16, 32, 72, 96, 128, 144, 152, 192, 384, 512]
MASKABLE_SIZES = [192, 512]

def hex_to_rgb(hex_color):
    """Converte colore hex in RGB tuple"""
    hex_color = hex_color.lstrip('#')
    return tuple(int(hex_color[i:i+2], 16) for i in (0, 2, 4))

def create_base_icon(size=512):
    """
    Crea icona base con logo QS Studio stilizzato
    
    Design: Cerchio blu con "Q" bianco (rappresenta Qualità)
    """
    # Crea immagine con sfondo trasparente
    img = Image.new('RGBA', (size, size), (255, 255, 255, 0))
    draw = ImageDraw.Draw(img)
    
    # Disegna cerchio di sfondo (primary color)
    padding = size // 8
    draw.ellipse(
        [padding, padding, size - padding, size - padding],
        fill=hex_to_rgb(PRIMARY_COLOR),
        outline=None
    )
    
    # Disegna lettera "Q" stilizzata (semplificata senza font)
    # Cerchio esterno Q
    q_padding = size // 3.5
    draw.ellipse(
        [q_padding, q_padding, size - q_padding, size - q_padding],
        fill=None,
        outline=hex_to_rgb(TEXT_COLOR),
        width=max(size // 20, 8)
    )
    
    # Buco interno Q
    q_hole_padding = size // 2.5
    draw.ellipse(
        [q_hole_padding, q_hole_padding, size - q_hole_padding, size - q_hole_padding],
        fill=hex_to_rgb(PRIMARY_COLOR),
        outline=None
    )
    
    # Coda Q (linea diagonale in basso a destra)
    tail_start_x = size * 0.65
    tail_start_y = size * 0.65
    tail_end_x = size * 0.78
    tail_end_y = size * 0.78
    draw.line(
        [(tail_start_x, tail_start_y), (tail_end_x, tail_end_y)],
        fill=hex_to_rgb(TEXT_COLOR),
        width=max(size // 20, 8)
    )
    
    return img

def create_maskable_icon(base_img, size):
    """
    Crea icona maskable (safe zone 80% del centro)
    Android può applicare maschere (cerchio, squircle, etc)
    """
    # Crea canvas più grande (125% per safe zone)
    canvas_size = int(size * 1.25)
    canvas = Image.new('RGBA', (canvas_size, canvas_size), hex_to_rgb(BACKGROUND_COLOR) + (255,))
    
    # Ridimensiona icona base
    resized = base_img.resize((size, size), Image.Resampling.LANCZOS)
    
    # Centra icona nel canvas
    offset = (canvas_size - size) // 2
    canvas.paste(resized, (offset, offset), resized)
    
    # Ridimensiona a dimensione target
    return canvas.resize((size, size), Image.Resampling.LANCZOS)

def generate_icons():
    """Genera tutte le icone richieste"""
    # Crea directory icons se non esiste
    icons_dir = os.path.join(os.path.dirname(__file__), 'icons')
    os.makedirs(icons_dir, exist_ok=True)
    
    print("🎨 Generazione icone PWA per SGQ ISO 9001:2015...")
    
    # Crea icona base ad alta risoluzione
    base_icon = create_base_icon(size=512)
    
    # Genera icone standard
    for size in SIZES:
        icon = base_icon.resize((size, size), Image.Resampling.LANCZOS)
        filename = f"icon-{size}x{size}.png"
        filepath = os.path.join(icons_dir, filename)
        icon.save(filepath, 'PNG', optimize=True)
        print(f"  ✅ {filename} ({size}x{size})")
    
    # Genera icone maskable
    for size in MASKABLE_SIZES:
        icon = create_maskable_icon(base_icon, size)
        filename = f"icon-maskable-{size}x{size}.png"
        filepath = os.path.join(icons_dir, filename)
        icon.save(filepath, 'PNG', optimize=True)
        print(f"  ✅ {filename} (maskable)")
    
    # Genera favicon.ico (multi-size)
    favicon_sizes = [(16, 16), (32, 32), (48, 48)]
    favicon_images = [base_icon.resize(size, Image.Resampling.LANCZOS) for size in favicon_sizes]
    favicon_path = os.path.join(icons_dir, '..', 'favicon.ico')
    favicon_images[0].save(
        favicon_path,
        format='ICO',
        sizes=favicon_sizes,
        append_images=favicon_images[1:]
    )
    print(f"  ✅ favicon.ico (multi-size)")
    
    print(f"\n✅ Icone generate in: {icons_dir}")
    print(f"📊 Totale: {len(SIZES)} standard + {len(MASKABLE_SIZES)} maskable + 1 favicon")

if __name__ == '__main__':
    try:
        generate_icons()
    except ImportError:
        print("❌ Errore: Pillow non installato")
        print("Installa con: pip install pillow")
        exit(1)
    except Exception as e:
        print(f"❌ Errore durante generazione icone: {e}")
        exit(1)

File: app/public/generate-icons.js
/**
 * Generatore Icone PWA - Sistema Gestione ISO 9001:2015
 * 
 * Genera icone placeholder SVG per PWA manifest.
 * Per icone professionali, sostituire con grafica custom.
 * 
 * Standard: ISO 9001:2015 punto 7.5 (documented information)
 * 
 * USO: node generate-icons.js
 */

const fs = require('fs');
const path = require('path');

const SIZES = [16, 32, 72, 96, 128, 144, 152, 192, 384, 512];
const MASKABLE_SIZES = [192, 512];
const PRIMARY_COLOR = '#1976d2';
const TEXT_COLOR = '#ffffff';

/**
 * Genera SVG con logo QS Studio stilizzato
 */
function generateSVG(size, isMaskable = false) {
  const viewBox = isMaskable ? size * 1.25 : size;
  const offset = isMaskable ? (viewBox - size) / 2 : 0;
  
  return `<?xml version="1.0" encoding="UTF-8"?>
<svg width="${size}" height="${size}" viewBox="0 0 ${viewBox} ${viewBox}" xmlns="http://www.w3.org/2000/svg">
  ${isMaskable ? `<rect width="${viewBox}" height="${viewBox}" fill="${PRIMARY_COLOR}"/>` : ''}
  <g transform="translate(${offset}, ${offset})">
    <!-- Cerchio sfondo -->
    <circle cx="${size/2}" cy="${size/2}" r="${size/2 - size/8}" fill="${PRIMARY_COLOR}"/>
    
    <!-- Lettera Q stilizzata -->
    <g>
      <!-- Cerchio esterno Q -->
      <circle cx="${size/2}" cy="${size/2}" r="${size/2 - size/3.5}" 
              fill="none" stroke="${TEXT_COLOR}" stroke-width="${size/20}"/>
      
      <!-- Buco interno Q -->
      <circle cx="${size/2}" cy="${size/2}" r="${size/2 - size/2.5}" fill="${PRIMARY_COLOR}"/>
      
      <!-- Coda Q -->
      <line x1="${size * 0.65}" y1="${size * 0.65}" 
            x2="${size * 0.78}" y2="${size * 0.78}" 
            stroke="${TEXT_COLOR}" stroke-width="${size/20}" stroke-linecap="round"/>
    </g>
    
    <!-- Testo SGQ (opzionale per icone grandi) -->
    ${size >= 192 ? `
    <text x="${size/2}" y="${size - size/6}" 
          text-anchor="middle" 
          fill="${TEXT_COLOR}" 
          font-family="Arial, sans-serif" 
          font-size="${size/12}" 
          font-weight="bold">SGQ</text>
    ` : ''}
  </g>
</svg>`;
}

/**
 * Salva SVG su file
 */
function saveSVG(size, isMaskable = false) {
  const iconsDir = path.join(__dirname, 'icons');
  
  // Crea directory se non esiste
  if (!fs.existsSync(iconsDir)) {
    fs.mkdirSync(iconsDir, { recursive: true });
  }
  
  const filename = isMaskable 
    ? `icon-maskable-${size}x${size}.svg`
    : `icon-${size}x${size}.svg`;
  
  const filepath = path.join(iconsDir, filename);
  const svgContent = generateSVG(size, isMaskable);
  
  fs.writeFileSync(filepath, svgContent, 'utf8');
  
  return filename;
}

/**
 * Crea favicon.svg (moderna alternativa a .ico)
 */
function createFavicon() {
  const faviconPath = path.join(__dirname, 'favicon.svg');
  const svg = generateSVG(32, false);
  fs.writeFileSync(faviconPath, svg, 'utf8');
  return 'favicon.svg';
}

/**
 * Crea README per icone
 */
function createIconsReadme() {
  const readme = `# Icone PWA - Sistema Gestione ISO 9001:2015

## Generazione

Icone SVG generate automaticamente da \`generate-icons.js\`.

## Conversione SVG → PNG (opzionale)

Per convertire SVG in PNG di alta qualità:

\`\`\`bash
# Con ImageMagick
for file in icons/*.svg; do
  convert -background none -density 300 "$file" "\${file%.svg}.png"
done

# Con Inkscape
for file in icons/*.svg; do
  inkscape "$file" --export-filename="\${file%.svg}.png" --export-dpi=300
done
\`\`\`

## Icone Custom

Per sostituire con icone professionali:
1. Creare PNG ad alta risoluzione (almeno 512x512)
2. Usare tool come [PWA Asset Generator](https://github.com/elegantapp/pwa-asset-generator)
3. Sovrascrivere file in questa directory

## File Generati

- \`icon-{size}x{size}.svg\` - Icone standard (${SIZES.join(', ')}px)
- \`icon-maskable-{size}x{size}.svg\` - Icone maskable per Android (${MASKABLE_SIZES.join(', ')}px)
- \`../favicon.svg\` - Favicon moderna (scalabile)

## Note Android

Icone **maskable** hanno safe zone 80% per compatibilità con maschere Android (cerchio, squircle, etc).
`;

  const readmePath = path.join(__dirname, 'icons', 'README.md');
  fs.writeFileSync(readmePath, readme, 'utf8');
}

/**
 * Main
 */
function main() {
  console.log('🎨 Generazione icone PWA per SGQ ISO 9001:2015...\n');
  
  let count = 0;
  
  // Genera icone standard
  for (const size of SIZES) {
    const filename = saveSVG(size, false);
    console.log(`  ✅ ${filename} (${size}x${size})`);
    count++;
  }
  
  // Genera icone maskable
  for (const size of MASKABLE_SIZES) {
    const filename = saveSVG(size, true);
    console.log(`  ✅ ${filename} (maskable)`);
    count++;
  }
  
  // Genera favicon
  const faviconName = createFavicon();
  console.log(`  ✅ ${faviconName} (favicon moderna)`);
  count++;
  
  // Crea README
  createIconsReadme();
  console.log(`  ✅ icons/README.md`);
  
  console.log(`\n✅ Icone generate: ${count} file`);
  console.log(`📁 Directory: ${path.join(__dirname, 'icons')}`);
  console.log('\n⚠️  NOTA: File SVG generati. Per PNG converti con ImageMagick/Inkscape.');
  console.log('   Oppure sostituisci con icone professionali custom.\n');
}

if (require.main === module) {
  main();
}

module.exports = { generateSVG, saveSVG };
